### ISSUE TITLE
optional parameters with default values and description or labels

### ISSUE THREAD

**Issue by shantanugadgil** [hashicorp/nomad#17204]
### Proposal

For `parameerized` jobs we can only specify the parameter name for the `required` or `optional` parameters.

How about if we could specify more information about the parameters in general like:
* a `default` value for `optional` parameters
* a `description` field, which would be used in the Nomad GUI (in the `Dispatch` section)
* this thought process could be enhanced further to specify the data type like string, number, etc.

### Use-cases
When an operator is using the Nomad GUI, I would like to have the "default" value of the optional parameter populated in the GUI.
Also, on a mouseover on the paramter, I would like to see the description which would help the user to fill in correct values.

### Attempted Solutions
N/A

**Comment by tgross** [comment]
Hi @shantanugadgil! This idea sounds interesting, can we dig into this a little further? Right now the `parameterized` block looks like this:

```hcl
parameterized {
    payload       = "required" # can also be "optional" or "forbidden"
    meta_required = ["first"]
    meta_optional = ["second", "third"]
  }
```

And then when we dispatch we do so with something like: `nomad job dispatch -meta first=1 second=2 third=3 example $payload_source`. So are you suggesting that we expand the `meta_required` and `meta_optional` fields here? Maybe we could merge the two blocks into a single `meta` block that has subblocks. The names here are terrible but something like:

```hcl
parameterized {
    payload       = "required" # can also be "optional" or "forbidden"

    metadata {

      meta "first" {
         required = true
         description = "the first thing"
      }

      meta "first" {
         required = false
         description = "the optional second thing"
         default = "example"
      }

    }
}
```

I'm not sure what you mean by the `label` field though. How would that work and what would it change to have it set? 

Having non-string types for metadata _might_ be possible but it gets hairy but then we need to store `any` types rather than the existing `meta`, and that'd have downstream effects when we start interpolating the metadata as [`meta`](https://developer.hashicorp.com/nomad/docs/job-specification/meta) blocks. Might be doable but something we'd need to discuss.

**Comment by shantanugadgil** [comment]
This is spot on. :+1: 

The change in structure is exactly what I had in mind, though I have never understood why it was "meta_" and not "param_"  (# nit-picking)


```hcl

parameterized {
  payload = "required" # can also be "optional" or "forbidden"

  parameters {           # <----  seems more natural to me, but I don't really care what it's called
    meta "first" {
      required    = true
      description = "the first thing"
      label       = "First Parameter"
      type        = "string"    # <---  wishlist, not super critical
    }

    meta "second" {
      required    = false
      description = "the optional second thing"
      default     = "example"
      label       = "Second Parameter"
      type        = "number"    # <---  wishlist, not super critical
    }
  }
}

```




> I'm not sure what you mean by the label field though. How would that work and what would it change to have it set?

The `label` is mostly a GUI thing which could be shown in place of `Start *` as shown in the below screenshot.

![image](https://github.com/hashicorp/nomad/assets/2508915/fef5980f-5764-4606-8c15-be321f897ba2)


About the non-string data types ... I think I was getting _carried away_ re-imagining the Dispatch UI in place of the Submit interface of a CI/CD tool like Jenkins.  _This is not essential to my use-case._

Wish:
I was also wondering if we could configure in the job definition, to make the input editbox be shown with a **spinner** control for "number" type of parameters (only allow increment/decrement)

_There are more things I could suggest from a UI perspective, but I think it would be good to limit the current scope only to the basic requirements of `default` ._  :smile:

**Comment by tgross** [comment]
> though I have never understood why it was "meta_" and not "param_"

Probably because it literally gets added to the job's `meta` and not passed as parameters to the task. It's up to the task to interpolate that however it sees fit (or not!).

> The label is mostly a GUI thing which could be shown in place of Start * as shown in the below screenshot.

Oh nice! I don't use the web UI enough, do I? :grinning: 

Ok, I'm going to mark this for roadmapping. If folks have other use cases, we'd love to hear about them!

**Comment by shantanugadgil** [comment]
follow up thought: Could the **required** parameters also have a "default" value?

**Comment by jmnavarrol** [comment]
...and, since we are at it, how about the ability of validate parameters out of a list?

I have batch jobs that would heavily benefit from  parameters' values selected from a list on the webui (and validated to belong to the list when job launched from command line)

**Comment by next-jesusmanuelnavarro** [comment]
Also, as a workaround for the description part, the ui block can be used now, i.e.:
```hcl2
	ui {
		description = <<-EOD
		**THIS IS A PARAMETERIZED DISPATCH**
		**Parameters:**
		* **FIRST_PARAMETER:** (mandatory) This sets FOO.
		* **OPTIONAL_PARAMETER:** (optional) Sets BAR to do something.

		More info about this job at [SOME URL](http://some.url/mydistpatch.html).
		EOD
	}
```

**Comment by jmnavarrol** [comment]
> Hi [@shantanugadgil](https://github.com/shantanugadgil)! This idea sounds interesting, can we dig into this a little further? 
> Having non-string types for metadata _might_ be possible but it gets hairy but then we need to store `any` types rather than the existing `meta`, and that'd have downstream effects when we start interpolating the metadata as [`meta`](https://developer.hashicorp.com/nomad/docs/job-specification/meta) blocks. Might be doable but something we'd need to discuss.

It might look interesting, but let's not forget that params map to environment variables into the running container which are always to be strings with any further validation needed to be done within the running environment itself.  So, either you control both the running environment and the jobspec, in which case this double checking may be redundant, or you only control one of them, in which case whatever the runtime does, it's the *"source of authority"*.  I mean, this is not a case where you can make use of all the validation capabilities that may HCL offer but a very limited subset of them.  Is it worth it, or would it be more confusing? (I suppose I'm defending a case of *"less is more"*)

I'd suggest tiering the development, from most valuable/easy to less:
1. Add the description field (as stated, this would mean parameters to move from a list to a list of dictionaries).  This would also allow for *"non-mandatory type hints"*, i.e.: *description: Please, insure this looks like an octal number"*.
2. Allow for the parameter "window" to become a list selector, ideally the list validation would also work on command line invocations.  This partially would also cover the "parameter type" scenario when the choices are limited.
3. I'm not so sure about the "label" concept: the parameter name already has to be a short string, thus acting already almost as a label, and the related environment variable can also be considered kind of a label too.
4. Same for validations: given a parameter's value **must** be a string and that it also should be validated by the runtime ¿does really add that much validating *"a string that looks like a string"* or *"a string that looks like a number"* or *"a blob that can be rendered to the shell as a string"*, etc?

**Comment by tgross** [comment]
Yeah at the end of the day the task driver is going to get a bunch of strings. Given all the above bits I sort of wonder if we could leverage the existing mechanisms we have for [HCL2 Variables](https://developer.hashicorp.com/nomad/docs/reference/hcl2/variables) to interpret the configuration of a dispatch meta block as well. That way there's not two different UXs for validation.

**Comment by jmnavarrol** [comment]
An example of how it may look an implementation from the outside  (expanding/modifiying @shantanugadgil example above and using a syntax as near to current one as possibly):

```hcl
job "docs" {
  parameterized {
    payload       = "required"
    meta_required = [
       # parameter can be defined as a string, for backwards compatibility
       "dispatcher_email",  # simple string, just as it's already the case
       # ...or it may be defined as a dictionary (new format; all keys are optional)
       "complex_parameter" = [
            name = "complex_parameter",  # defaults to the parameter's key, so maybe unneeded
            description = "this parameter sets foo; select one of the values from the list",
            # since this parameter is mandatory, one of the values below should be chosen.
            # easiest approach for a default value should be always the empty string
            # but it may be worth to explore the possibility of using either a decorator or a "default" key
            # nomad should validate that the parameter's value is in fact one from the list below at run time
            value_list = [
                 "one",  # since values will be turned into environment variables, these should be checked as valid
                 "two",
                 "three",
            ],
            # default value optionally set using a property (if value_list is provided, it must be one of the values above
            # if not set, default is the empty string
            default_value = "two",
            # for a default using a decoration, see example below
       ],
    ]
    meta_optional = [
        # old-style parameter definition for backwards compatibility (a simple string)
        "pager_email",
        # new style
       "complex_optional_parameter" = [
       # just as above, each key here is optional; see how description may use markdown formatting
            name = "complex_optional_parameter",  # defaults to the parameter's key
            description = "this optionally sets bar; either empty or **one of the values below**"
            value_list = [
            # it might be worthy to define some kind of decoration to set a default
            # if decorator not present, default to the empty string
                "foo",
                @"bar",  # '@' decorator used here to signify the default value
                "zoot",
            ]
       ]
    ]
  }
}

```
